[PT1.11] make static graph to be stable

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71483
* __->__ #71459

1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

zhaojuanmao committed:
[PT1.11] make static graph to be stable

1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[PT1.11] make static graph to be stable"


1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
+                     will not change during the whole training loop (meaning there is
+                     no control flow depending on iterations). When graph is set to be
+                     static, DDP will support cases that can not be supported in the past:
+                     1) reentrant backwards 2) activation checkpointing multiple times
```
nit: capitalize after each number

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
+                     3) activation checkpointing with find_unused_parameters = true.
+                     4) not all output tensors are used in loss calculation. 5) there is
+                     model parameter that is outside of forward function. 6) potentially
+                     improve performance when find_unsued_parameters = true or there are
```
nit: True

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
+                     1) reentrant backwards 2) activation checkpointing multiple times
+                     3) activation checkpointing with find_unused_parameters = true.
+                     4) not all output tensors are used in loss calculation. 5) there is
+                     model parameter that is outside of forward function. 6) potentially
```
nit: Seems like we have period before certain numbers and not others.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -1635,30 +1654,23 @@ def _set_ddp_runtime_logging_sample_rate(self, sample_rate):
-        This API should be called after DistributedDataParallel construction, and
-        before training loops starts. Also it should be called in the same way for
-        all ranks. For example:
+        It is recommended to set static graph in the DDP constructor, which will
```
Maybe just remove docs all together for this since it is not a public API we want to support for end users?

zhaojuanmao replied:
sounds good

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -727,7 +745,8 @@ def __setstate__(self, state):
         self._ddp_init_helper(parameters, expect_sparse_gradient, param_to_name_mapping)
         if self.static_graph:
-            self._set_static_graph()
+            self.reducer._set_static_graph()
```
Why can't we just call `self._set_static_graph()` here?

zhaojuanmao replied:
since I let _set_static_graph() early return if self.static_graph=True, to make it work even users set static_graph=True and accidentally call _set_static_graph() again.  

because of this change, calling self. _set_static_graph() will early return, not really doing anything here

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
+                     no control flow depending on iterations). When graph is set to be
+                     static, DDP will support cases that can not be supported in the past:
+                     1) reentrant backwards 2) activation checkpointing multiple times
+                     3) activation checkpointing with find_unused_parameters = true.
```
Since we earlier in the doc say that find_Unused_params=True doesn't matter when static graph is set, maybe we can clarify this as "activation checkpointing when model has unused parameters".

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
+                     static, DDP will support cases that can not be supported in the past:
+                     1) reentrant backwards 2) activation checkpointing multiple times
+                     3) activation checkpointing with find_unused_parameters = true.
+                     4) not all output tensors are used in loss calculation. 5) there is
```
(4) is now supported in general, so maybe we can omit this?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -478,6 +478,21 @@ class DistributedDataParallel(Module, Joinable):
                       gradients. If hitting such errors, please fix it by
                       referring to the :meth:`~torch.optim.Optimizer.zero_grad`
                       function in ``torch/optim/optimizer.py`` as a solution.
+        static_graph (bool): When set to ``True``, DDP knows the trained graph is
```
It might be helpful for users to understand how to figure out if their graph is static or not (is there an easy check to do this)? The main reason is that sometimes model authoring code is very separated from model training code, so it might not always be completely straightforward.

zhaojuanmao replied:
that is a good point! let me add a github issue about this

zhaojuanmao replied:
added an example about how to check this in this PR now

rohan-varma reviewed (approved):
LGTM, stamping to unblock. Please also address @pritamdamania87 's comments

zhaojuanmao committed:
Update on "[PT1.11] make static graph to be stable"


1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[PT1.11] make static graph to be stable"


1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[PT1.11] make static graph to be stable"


1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[PT1.11] make static graph to be stable"


1. add static_graph feature to DDP constructor;
2. still keep _set_static_graph() API, so that existing use cases are not affected, also it can be called internally by DDP constructor
3. four cases are covered:
    static_graph = False, _set_static_graph() is called;
    static_graph = False, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is not called;
    static_graph = True, _set_static_graph() is called;

Differential Revision: [D33646738](https://our.internmc.facebook.com/intern/diff/D33646738/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
